examples/control-vector-generator/pca.hpp

```diff
@@ -0,0 +1,337 @@
+
+    // delete the temporally context used to build the graph
+    ggml_free(ctx0);
+    return gf;
+}
+
+static ggml_status compute_piter(
+        const struct pca_params & params,
+        const pca_model & model,
+        struct ggml_cgraph * gf,
+        ggml_gallocr_t allocr,
+        struct pca_result & result) {
+    // allocate tensors
+    ggml_gallocr_alloc_graph(allocr, gf);
+
+    if (ggml_backend_is_cpu(model.backend)) {
+        ggml_backend_cpu_set_n_threads(model.backend, params.n_threads);
+    }
+
+#ifdef GGML_USE_METAL
+    if (ggml_backend_is_metal(model.backend)) {
+        ggml_backend_metal_set_n_cb(model.backend, params.n_threads);
+    }
+#endif
+
+    ggml_status res = ggml_backend_graph_compute(model.backend, gf);
+    if (res == GGML_STATUS_SUCCESS) {
+        auto extract_i = [](std::string prefix, std::string str) -> int {
+            int i = -1;
+            if (str.rfind(prefix, 0) == 0) {
+                sscanf(str.c_str(), (prefix + "%d").c_str(), &i);
+            }
+            return i;
+        };
+        result.calculated_square = NULL;
+        result.eigenvectors.clear();
+        result.distances.clear();
+        result.eigenvectors.resize(params.n_batch);
+        result.distances.resize(params.n_batch);
+        // get output nodes
+        for (int i = 0; i < gf->n_nodes; ++i) {
+            auto node = gf->nodes[i];
+            int iter = -1;
+            // find b_tensor (without copying data from device)
+            if ((iter = extract_i("b_tensor_norm_", node->name)) > -1) {
+                result.eigenvectors[iter] = node;
+            }
+            // find distances, then copy data from device
+            if ((iter = extract_i("distance_", node->name)) > -1) {
+                float d;
+                ggml_backend_tensor_get(node, &d, 0, sizeof(float));
+                result.distances[iter] = d;
+                // std::cout << node->name << " = " << d << "\n";
+            }
+            // find tmp_square if it exists (without copying data from device)
+            if (std::string(node->name) == "tmp_square") {
+                result.calculated_square = node;
+            }
+        }
+    }
+    return res;
+}
+
+static void power_iteration(
+        const struct pca_params & params,
+        struct ggml_tensor * input, // shape of input: [n_samples, n_embd]
+        struct ggml_tensor * output) {
+    //printf("in power iteration\n");
+    struct pca_model model(input);
+
+    ggml_gallocr_t allocr = NULL;
+    struct pca_result result;
+    struct ggml_tensor * last_eigenvector = NULL;
+
+    int n_iters = params.n_iterations / params.n_batch; // more batch, fewer iterations
+    for (int iter = 0; iter < n_iters; ++iter) {
+        bool calc_square = (iter == 0); // only need to calculate square for first iteration
+        if (allocr) {
+            ggml_gallocr_free(allocr);
+        }
+        allocr = ggml_gallocr_new(ggml_backend_get_default_buffer_type(model.backend));
```

Comment:
Unless I am missing something about the logic here, it is not necessary to create a new gallocr. You can reuse the same one to allocate any graphs, as long as the buffer type is the same.

Reply:
Yeah I was also unsure about this point. My question is: Will re-using the same `allocr` also "free" the old `gf`? For example:

```cpp
ggml_gallocr_t allocr = ggml_gallocr_new(...);

struct ggml_cgraph * gf0 = build_graph_piter(...);
ggml_gallocr_alloc_graph(allocr, gf0);
ggml_backend_graph_compute(model.backend, gf0);

struct ggml_cgraph * gf1 = build_graph_piter(...);
ggml_gallocr_alloc_graph(allocr, gf1); // reuse allocr, will it automatically "free" gf0?
ggml_backend_graph_compute(model.backend, gf1);
```

Reply:
Yes, something like that. It will reuse the same buffer to allocate all the graphs. However the buffer may need to be reallocated, which will cause the previously allocated graphs to become invalid. So generally you should only use the last graph allocated by a gallocr.

Reply:
Thanks for the explanation. It's safe to reuse allocr then.

I removed this code in https://github.com/ggerganov/llama.cpp/pull/7514/commits/f54cb8e3075832dcb0edfa4fcd7a2e01522bf0dd